### PR TITLE
DM-35591 Prototype AnalysisTools

### PR DESCRIPTION
**Pull request by natelust**

### CODE REVIEW

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    KeyedDataSchema,
+    Vector,
+)
+from ..scalarActions import CountAction, MedianAction, SigmaMadAction
+from ..keyedDataActions import AddComputedVector, KeyedScalars
+from ..vectorActions import SnSelector, StellarSelector, FractionalDifference
+from ..vectorActions.calcShapeSize import CalcShapeSize
+
+
+class _AproxMedian(ScalarAction):
```
Comment (bechtol) [inline python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:28]: Should this class be moved to scalarActions/scalarActions.py ?

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    vectorKey = Field[str](doc="Key for the vector to perform action on", optional=False)
+
+    def getInputSchema(self) -> KeyedDataSchema:
+        return ((self.vectorKey, Vector),)  # type: ignore
+
+    def __call__(self, data: KeyedData, **kwargs) -> Scalar:
+        mask = self.getMask(**kwargs)
+        value = np.sort(data[self.vectorKey.format(**kwargs)][mask])  # type: ignore
+        x = int(len(value) / 10)
+        return np.nanmedian(value[-x:])
```
Comment (bechtol) [inline python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:38]: I think this should be 

return np.nanmedian(value[::x])

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py (hunk 1 above)
Comment (laurenam) [reply python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:28]: Also, is "Aprox" here meant to be short for "Approximately"?  If so, check your "p"s (and if not, perhaps a full spell-out would avoid confusion?)

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    snFluxType = Field[str](doc="column key for the flux type used in SN selection")
+
+    selector = ConfigurableActionField[VectorAction](doc="Selector to use before computing Scalars")
+
+    def setDefaults(self):
+        super().setDefaults()
+        self.scalarActions.median = MedianAction(colKey=self.columnKey)  # type: ignore
+        self.scalarActions.sigmaMad = SigmaMadAction(colKey=self.columnKey)  # type: ignore
+        self.scalarActions.count = CountAction(colKey=self.columnKey)  # type: ignore
+        self.scalarActions.approxMag = _AproxMedian(  # type: ignore
```
Comment (laurenam) [inline python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:53]: Mind your "p"s (but worry not about any "q"s) 😉

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py (hunk 1 above)
Comment (natelust) [reply python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:28]: This was intended to be something specific implementation detail specific for how this calculates something, and may get replaces with something better later. I have no strong feeling if others really want to move it I will not argue, but if others don't have strong feelings I would prefer to leave it where it is.

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py (hunk 2 above)
Comment (natelust) [reply python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:38]: No, in this case it is doing exactly what it should. it sorts the array, finds points that are considered the "tail" and only wants to find the median of those tail values.

This is used as a heuristic to determine approximately what magnitude corresponds to a certain signal to noise.

File: python/lsst/analysis/tools/scalarActions/scalarActions.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    vectorKey = Field[str](doc="Key for the vector to perform action on", optional=False)
+
+    def getInputSchema(self, **kwargs) -> KeyedDataSchema:
+        return ((self.vectorKey.format(**kwargs), Vector),)
+
+    def __call__(self, data: KeyedData, **kwargs) -> Scalar:
+        mask = self.getMask(**kwargs)
+        value = np.sort(data[self.vectorKey.format(**kwargs)][mask])  # type: ignore
+        x = int(len(value) / 10)
+        return np.nanmedian(value[-x:])
```
[the remark this answers is not in the document]
Comment (natelust) [reply python/lsst/analysis/tools/scalarActions/scalarActions.py:79]: ah in contrast to what was above, I recall talking to Sophie about her wanting it else where so perhaps we need to consolidate versions. See other comments for discussion on internals

File: python/lsst/analysis/tools/analysisMetrics/analysisMetrics.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        self.prep.selectors.starSelector = StellarSelector()
+
+        self.process.buildActions.x = ExtinctionCorrectedMagDiff()
+        self.process.buildActions.x.magDiff.returnMillimags = False
+        self.process.buildActions.y = ExtinctionCorrectedMagDiff()
+        self.process.buildActions.y.magDiff.returnMillimags = False
+
+
+class GRIStellarPSFMetric(AnalysisMetric):
+    # Use this as the Base Class for now StellarLocusBaseMetric
```
Comment (bechtol) [inline python/lsst/analysis/tools/analysisMetrics/analysisMetrics.py:94]: This comment seems inconsistent with the class inheritance pattern that is being used.

File: python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py (hunk 2 above)
Comment (bechtol) [reply python/lsst/analysis/tools/analysisParts/shapeSizeFractional.py:37]: Oh, I was thrown off by the name of the class suggesting a median

**Pull request merged**